### PR TITLE
POC/Provisioning: update spec and examples

### PR DESCRIPTION
**Pull request by ryantxu**
This modifies the spec in a few ways:
1. adds title + description
2. adds an explicit "type" rather than detect it
3. Uses the suffix "Config" for each of the typed configs -- this will leave the non-config name avaliable when we need a real implementation!

```
❯ kubectl apply -f pkg/tests/apis/provisioning/testdata/github-example.yaml
repository.provisioning.grafana.app/github-example created
❯ kubectl apply -f pkg/tests/apis/provisioning/testdata/s3-example.yaml
repository.provisioning.grafana.app/s3-example created
❯ kubectl apply -f pkg/tests/apis/provisioning/testdata/local-devenv.yaml
repository.provisioning.grafana.app/local-devenv created
❯ kubectl get repositories
NAME             CREATED AT             TITLE            TYPE     TARGET
github-example   2024-11-14T09:20:00Z   Github Example   github   grafana/git-ui-sync-demo
local-devenv     2024-11-14T09:20:12Z   S3 Example       local    path/to/folder
s3-example       2024-11-14T09:20:05Z   S3 Example       s3       my-bucket
```

image

### CODE REVIEW

File: pkg/registry/apis/provisioning/hello.go [hunk 1]
```diff
@@ -42,13 +47,27 @@ func (*helloWorldSubresource) NewConnectOptions() (runtime.Object, bool, string)
...
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		whom := r.URL.Query().Get("whom")
 		if whom == "" {
 			whom = "World"
 		}
 
+		fmt.Printf("GOT: %s/%s\n", repo.Name, repo.Spec.Type)
+		fmt.Printf(" local: %+v\n", repo.Spec.Local)
+		fmt.Printf(" github: %+v\n", repo.Spec.GitHub)
+		fmt.Printf(" s3: %+v\n", repo.Spec.S3)
```
Comment (ryantxu) [inline pkg/registry/apis/provisioning/hello.go:69]: @Proximyst -- this seems to keep the nils as expected:

for http://localhost:3000/apis/provisioning.grafana.app/v0alpha1/namespaces/default/repositories/github-example/hello

I see:
```
GOT: github-example/github
 local: <nil>
 github: &{Owner:grafana Repository:git-ui-sync-demo BranchWorkflow:true GenerateDashboardPreviews:true}
 s3: <nil>
```

File: pkg/apis/provisioning/v0alpha1/types.go [hunk 2]
```diff
@@ -56,50 +45,43 @@ type GitHubRepository struct {
...
-		if !it.IsEmpty() {
-			nonEmpty++
-		}
-	}
-	if nonEmpty != 1 {
-		return fmt.Errorf("%w (found %d)", errors.New("one (and exactly one) of github, s3, and local must be set"), nonEmpty)
-	}
-
-	return nil
+	GitHub *GitHubRepositoryConfig `json:"github,omitempty"`
```
Comment (Clarity-89) [inline pkg/apis/provisioning/v0alpha1/types.go:84]: Are we specifically targeting GItHub or Git in general?

Comment (ryantxu) [reply pkg/apis/provisioning/v0alpha1/types.go:84]: Eventually we will have both -- but github alone will have specific config/features that just `git` would not.

Comment (ryantxu) [reply pkg/apis/provisioning/v0alpha1/types.go:84]: eg -- webhooks and posting previews to github comments is speciffically github, but listing files could be any git

File: pkg/apis/provisioning/v0alpha1/types.go [hunk 3]
```diff
@@ -56,50 +45,43 @@ type GitHubRepository struct {
 	GenerateDashboardPreviews bool `json:"generateDashboardPreviews,omitempty"`
 }
 
-func (r *GitHubRepository) IsEmpty() bool {
-	// we don't need to check options here, just the most important stuff to actually connect.
-	return r == nil || r.Owner == "" || r.Repository == ""
-}
+// RepositoryType defines the types of Repository
+// +enum
```
Comment (Proximyst) [inline pkg/apis/provisioning/v0alpha1/types.go:49]: for my own learning's sake: what tool depends on this tag? openapi-gen?

**Review by Proximyst — APPROVED** [review]

File: pkg/apis/provisioning/v0alpha1/types.go (hunk 3 above)
Comment (ryantxu) [reply pkg/apis/provisioning/v0alpha1/types.go:49]: kube-openapi looks for it https://github.com/kubernetes/kube-openapi/blob/master/pkg/generators/enum.go#L168

and then will set the enum values in openapi:
https://github.com/grafana/grafana/pull/96438/files#diff-2d0d891a8c382777f4d12dd64bb2c91b5cec99cf3245631e99f4b4f6c560e12dR238

i don't think that is yet supported in the app-platform cue > CRD codegen

**Pull request merged**